src/test/scala/com/microsoft/hyperspace/index/DeltaLakeIntegrationTest.scala

```diff
@@ -242,10 +242,12 @@ class DeltaLakeIntegrationTest extends QueryTest with HyperspaceSuite {
             spark.disableHyperspace()
             val dfWithHyperspaceDisabled = query()
             val basePlan = dfWithHyperspaceDisabled.queryExecution.optimizedPlan
+            val resultDisabled = dfWithHyperspaceDisabled.collect().toSeq.toSet
             spark.enableHyperspace()
             val dfWithHyperspaceEnabled = query()
             assert(!basePlan.equals(dfWithHyperspaceEnabled.queryExecution.optimizedPlan))
-            checkAnswer(dfWithHyperspaceDisabled, dfWithHyperspaceEnabled)
+            val resultEnabled = dfWithHyperspaceEnabled.collect().toSeq.toSet
+            assert(resultEnabled.equals(resultDisabled))
```

Comment:
is this change still needed? if so, why?

Reply:
In checkAnswer, a new thread tries to build the query plan again - so it causes the exception.
No idea why other `checkAnswer`s have no problem with it..

Reply:
oh interesting. could you share the code where a new thread is spawned in `checkAnswer`?

Reply:
It's because of broadcast join..

```
[info]          at org.apache.spark.sql.execution.exchange.BroadcastExchangeExec.doExecuteBroadcast(BroadcastExchangeExec.scala:146)
[info]          at org.apache.spark.sql.execution.InputAdapter.doExecuteBroadcast(WholeStageCodegenExec.scala:387)
[info]          at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeBroadcast$1(SparkPlan.scala:144)
[info]          at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:155)
[info]          at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
[info]          at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
[info]          at org.apache.spark.sql.execution.SparkPlan.executeBroadcast(SparkPlan.scala:140)
[info]          at org.apache.spark.sql.execution.joins.BroadcastHashJoinExec.prepareBroadcast(BroadcastHashJoinExec.scala:117)
[info]          at org.apache.spark.sql.execution.joins.BroadcastHashJoinExec.codegenInner(BroadcastHashJoinExec.scala:211)
[info]          at org.apache.spark.sql.execution.joins.BroadcastHashJoinExec.doConsume(BroadcastHashJoinExec.scala:101)
```

failed thread stack:
```
..
[info]          at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDown(LogicalPlan.scala:29)
[info]          at com.microsoft.hyperspace.index.rules.FilterIndexRule$.apply(FilterIndexRule.scala:52)
[info]          at com.microsoft.hyperspace.index.rules.FilterIndexRule$.apply(FilterIndexRule.scala:38)
[info]          at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$2(RuleExecutor.scala:87)
[info]          at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
[info]          at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
[info]          at scala.collection.immutable.List.foldLeft(List.scala:89)
[info]          at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$1(RuleExecutor.scala:84)
[info]          at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$1$adapted(RuleExecutor.scala:76)
[info]          at scala.collection.immutable.List.foreach(List.scala:392)
[info]          at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:76)
[info]          at org.apache.spark.sql.execution.QueryExecution.optimizedPlan$lzycompute(QueryExecution.scala:66)
[info]          at org.apache.spark.sql.execution.QueryExecution.optimizedPlan(QueryExecution.scala:66)
[info]          at org.apache.spark.sql.execution.QueryExecution.sparkPlan$lzycompute(QueryExecution.scala:72)
[info]          at org.apache.spark.sql.execution.QueryExecution.sparkPlan(QueryExecution.scala:68)
[info]          at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:77)
[info]          at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:77)
[info]          at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3359)
[info]          at org.apache.spark.sql.Dataset.collect(Dataset.scala:2782)
[info]          at org.apache.spark.sql.delta.PartitionFiltering.filesForScan(PartitionFiltering.scala:40)
[info]          at org.apache.spark.sql.delta.PartitionFiltering.filesForScan$(PartitionFiltering.scala:27)
[info]          at org.apache.spark.sql.delta.Snapshot.filesForScan(Snapshot.scala:52)
[info]          at org.apache.spark.sql.delta.files.TahoeLogFileIndex.matchingFiles(TahoeFileIndex.scala:140)
[info]          at org.apache.spark.sql.delta.files.TahoeFileIndex.listFiles(TahoeFileIndex.scala:56)
[info]          at org.apache.spark.sql.execution.FileSourceScanExec.selectedPartitions$lzycompute(DataSourceScanExec.scala:193)
[info]          at org.apache.spark.sql.execution.FileSourceScanExec.selectedPartitions(DataSourceScanExec.scala:190)
[info]          at org.apache.spark.sql.execution.FileSourceScanExec.updateDriverMetrics(DataSourceScanExec.scala:529)
[info]          at org.apache.spark.sql.execution.FileSourceScanExec.inputRDD$lzycompute(DataSourceScanExec.scala:307)
[info]          at org.apache.spark.sql.execution.FileSourceScanExec.inputRDD(DataSourceScanExec.scala:305)
[info]          at org.apache.spark.sql.execution.FileSourceScanExec.inputRDDs(DataSourceScanExec.scala:327)
[info]          at org.apache.spark.sql.execution.FilterExec.inputRDDs(basicPhysicalOperators.scala:121)
[info]          at org.apache.spark.sql.execution.ProjectExec.inputRDDs(basicPhysicalOperators.scala:41)
[info]          at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:627)
```

Reply:
Seems we shouldn't throw the exception .. ?

Reply:
Hmm, then we have to think about this feature again, since it may hit into thread-safety issue even if the user didn't intend it?

Reply:
For example, this could have gone unnoticed if we didn't throw the exception?

Reply:
Seems so. How about ThreadLocal[HashMap[Session, Context]]?

Reply:
But that would allow multiple threads share the same Hyperspace context.

Reply:
So, basically in Delta code, it does `allFiles.toDF()` and eventually that dataframe is `collect`-ed, which triggers our optimizer. In our optimizer, we have an extractor that calls `Hyperspace.getContext(spark).sourceProviderManager` in `ExtractRelation`. So instead of `case filter @ Filter(condition: Expression, ExtractRelation(relation))`, we could do `case filter @ Filter if isSupportedRelation`, and I think we can avoid the issue.

But, I think this is still hacky and the right way to fix it seems like making the context thread-safe.